### DISCUSSION TITLE
API and Technical Documentation

### DISCUSSION THREAD

**Discussion by osamaeshtiaq** [saleor/saleor#7539] [General]
### What I'm trying to achieve
To understand the backend structure which is turning out to be extremely vague due to lack of API and technical documentation.

### Describe a proposed solution
A library can be used to integrate with graphene and generate API documentation.

**Comment by krzysztofwolski** [comment]
@osamaeshtiaq API documentation is already documented:
https://docs.saleor.io/docs/3.0/developer/api-reference/api-introduction

**Comment by onecrayon** [comment]
I completely agree with the OP; the GraphQL schema is indeed documented (to some extent), but there is very little documentation describing the underlying model relationships and logic. For instance, I needed to create a product through GraphQL. I can infer some things (like the fact that a ProductType ID is required means I need to create a ProductType first), but there's nothing outlining the process, which involves something like 4-8 distinct GraphQL calls. I had to create products by hand in the dashboard, monitor the network requests in my browser's inspector, and extrapolate from there. Things like the fact that you need to add a Product to channels via one GraphQL mutation, and then later assign prices to the ProductVariant through another are non-obvious because the underlying relationships are often opaque.

Basically, the docs do an alright job of describing how to query an already-populated store, which is good for our eventual front-end developers. The GraphQL automatic schema documentation is awesome once you know what's going on under the hood (though a lot of the descriptions are terse to the point of uselessness), but gaining that knowledge is currently reliant on trial and error, which is frustrating and time consuming.

**Reply by patrys** [reply]
That's all fair and on point. We keep improving the docs, and we'll get there eventually, but it takes time. So if you have a minute to spare, please consider submitting a small pull request. Even an outline would be a good starting point.

**Reply by onecrayon** [reply]
Well, that's kind of the problem; I don't have the high-level understanding of the models and flows to be able to sketch out an outline even. 😄 

At this point I'm doing a lot of trial and error testing through GraphQL to try and figure out requirements and how things flow.

**Comment by mirekm** [comment]
👋 Thank you @onecrayon, it's pretty spot on! We plan to dedicate way more time and resources to our documentation any moment soon.

1. Would you mind sharing the most pressing topics we should cover from your perspective so we can include it during our next documentation workshop?

2. Would you be interested in participating in a bigger community workshop/hackathon dedicated to docs?

**Reply by onecrayon** [reply]
@mirekm 

> Would you mind sharing the most pressing topics we should cover from your perspective

Checkout and Orders, hands-down. There are so many mutations, models, and hooks associated with these, and they're so central to the system, but I've been unable to find documentation about how the various models relate to one another, what is required to move from one step to another (e.g. what are the required pieces before you can take an Order from a draft to unfulfilled?), and both what the flow is *intended* to be normally versus what it is *allowed* to be (e.g. it would be useful to see both what the intended flow is for a credit card, but also know where/if validation restricts the flow when designing other payment approaches).

From a backend developer perspective, it would also be very useful to have all of the plug-in methods documented with the order they are called, what mutations/actions lead to them being called, and what is expected for the `previous_value`. Browsing through `base_plugin.py` give a high-level overview of what you can do, but the relationships between calls and higher-level flows aren't necessarily explicit, while `previous_value` is often completely opaque.

> Would you be interested in participating in a bigger community workshop/hackathon dedicated to docs?

I'm certainly willing to contribute, though I'm somewhat limited in that I don't actually understand the underlying logic all that well yet. 😄 I could potentially help move from outlines to full documentation.
